spec.html

```diff
@@ -21281,6 +21297,12 @@ <h1>Static Semantics: AllPrivateIdentifiersValid</h1>
         1. Append to _newNames_ the elements of PrivateBoundIdentifiers of |ClassBody|.
         1. Return AllPrivateIdentifiersValid of |ClassElementList| with argument _newNames_.
       </emu-alg>
+
+      <emu-grammar>RelationalExpression : PrivateIdentifier `in` ShiftExpression</emu-grammar>
+      <emu-alg>
+        1. If StringValue of |PrivateIdentifier| is in _names_, return *true*.
```

Comment:
I think I made a mistake here when I originally suggested this. This sees a an expression `#foo in SHIFT_EXPRESSION`, and decides that if `foo` is a known private, all is good. Except, we forgot to check whether there are any unknown privates in `SHIFT_EXPRESSION`. For instance, I don't believe we fail like we should with `#foo in obj.#bar` (given `#foo` is declared and `#bar` is not), because `#bar` is in the `SHIFT_EXPRESSION` position and we stop at the `RELATIONAL_EXPRESSION` position.

We can see that we still need to recurse by looking at the other `AllPrivateIdentifiersValid` defined here:

```text
MemberExpression : MemberExpression . PrivateIdentifier
  1. If names contains the StringValue of PrivateIdentifier, then
    a. Return AllPrivateIdentifiersValid of MemberExpression with argument names.
  2. Return false.
```

Here, we found `MEMBER_EXPRESION.#foo`, `foo` was a known private and then continue to recurse into `MEMBER_EXPRESSION`. Because there could be some `#bar` deeper down (eg, `obj.#bar.#foo`) that is not declared.

```suggestion
        1. If StringValue of |PrivateIdentifier| is in _names_, then
          1. Return AllPrivateIdentifiersValid of |ShiftExpression| with argument _names_.
```

Reply:
Agreed. #1668 originally had the same issue, in the other parts of AllPrivateIdentifiersValid.

While you're at it:

```suggestion
        1. If _names_ contains the StringValue of |PrivateIdentifier|, then
```

Reply:
Thanks! Both done.